### ISSUE TITLE
mitmproxy weird behaviour

### ISSUE THREAD

**Issue by shrikantkaramkar** [mitmproxy/mitmproxy#1414]
##### Steps to reproduce the problem:
```
1. Redirecting multiple request in my app.
2. all my redirections are: pointing server URL to local folders having stored data in files
3. When I start the mitmproxy with all redirection then as per code url's are redirecting to expected 
4. the problem I am facing is application is not loading, and to load the application I need to stop the mitmproxy.
5. Why should I need to stop the MITMPROXY to load the application.
```
`For example`
```
if flow.request.pretty_url.endswith("prompt=login&ui_locales=en"):

            print("In First Request If: ")
            resp = HTTPResponse(
                                 "HTTP/1.1",
                                 302,
                                 "Moved Temporarily",
                                 Headers([
                                            ["Server", "Apache"],
                                            ["Date", "Wed, 21 Jul 2016 23:23:48 GMT"],
                                            ["Location", "https://login.xxxxx.net/oauth/authorize"],
                                            ["Set-Cookie", "abc"],
                                            ["Vary", "Accept-Encoding"],
                                            ["Content-Encoding", "gzip"],
                                            ["Keep-Alive", "timeout=5, max=448"],
                                            ["Connection", "Keep-Alive"],
                                            ["Transfer-Encoding", "chunked"],
                                            ["Content-Type", "text/html"]
                                ]),
                                ("<html><head><title>302 Moved Temporarily</title></head>"
"<body bgcolor="#FFFFFF">"
"<p>This document you requested has moved temporarily.</p>"
"<p>It's now at <a href=\"https://login.xxxxxx.net/oauth/authorize</a>.</p>"
"</body></html>")
            )
            flow.reply(resp)
 @mhils
 mitmproxy member

```

`Mitmproxy Version: 0.15`
`Operating System: OSX`

**Comment by mhils** [comment]
How are you redirecting the requests?

**Comment by shrikantkaramkar** [comment]
I am using request_redirect.py for all my redirection.

**Comment by mhils** [comment]
redirect_requests.py shows two different ways. Can you post your code? I can't really reproduce what you are describing, sorry.

**Comment by shrikantkaramkar** [comment]
updated above.

**Comment by mhils** [comment]
I can't really reproduce this issue, sorry.

**Comment by joshuabuildsthings** [comment]
@shrikantkaramkar Your HTTP Response from mitmproxy lacks a Content-Length header. Please try adding one with the correct length and let me know if it fixes your problem.

**Comment by shrikantkaramkar** [comment]
I am getting below error on adding "Content-Length" in header
```
mitmproxy has crashed!
Please lodge a bug report at: https://github.com/mitmproxy/mitmproxy
10.0.0.62:58218: Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/mitmproxy/proxy/server.py", line 121, in handle
    root_layer()
  File "/Library/Python/2.7/site-packages/mitmproxy/proxy/modes/http_proxy.py", line 11, in __call__
    layer()
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/tls.py", line 358, in __call__
    layer()
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/http1.py", line 67, in __call__
    layer()
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/http.py", line 171, in __call__
    self.handle_regular_mode_connect(request)
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/http.py", line 263, in handle_regular_mode_connect
    layer()
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/tls.py", line 358, in __call__
    layer()
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/http1.py", line 67, in __call__
    layer()
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/http.py", line 200, in __call__
    self.log("response", "debug", [repr(flow.response)])
AttributeError: 'HTTPResponse' object has no attribute 'response'
```

**Comment by joshuabuildsthings** [comment]
@shrikantkaramkar 
The source of the crash from the Traceback you posted seems to be from within mitmproxy, but it's hard to know without more context. Can you please post the current code you're using that generates that error?

Thanks!

**Comment by shrikantkaramkar** [comment]
```
from mitmproxy.models import HTTPResponse
from netlib.http import Headers
import xml.etree.ElementTree as ET
import json
import time

mapping_json = ""

def start(context, argv):
    if len(argv) == 0:
        raise ValueError('Usage: -s "<script_name>.py mapping_json"')

    global mapping_json
    global mapped_data

    mapping_json = "".join(argv[1])
    print("Mapping JSON:", mapping_json)
    mapped_data = json.loads(mapping_json)
    print("In Start")

    if '/account/test/hi' in mapped_data:
        print("Found")

#-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------#

def request(context, flow):
        print("In request")
        global mapping_json
        if flow.request.pretty_url.endswith("prompt=login&ui_locales=en"):

                resp = HTTPResponse(
                                     "HTTP/1.1",
                                     302,
                                     "Moved Temporarily",
                                     Headers([
                                                ["Server", "Apache"],
                                                ["Date", "Wed, 21 Jul 2016 23:23:48 GMT"],
                                                ["Location", "https://login.abcd.net/oauth/authorize?%2A"],
                                                ["Set-Cookie", "bid=Ld9NnIJsRSKQF9o1HgVDJCGhka8; expires=Tue, 22-Oct-2016 19:18:48 GMT; secure,"],
                                                ["Set-Cookie", ("tg_ticket=VEdUAQEAABa1zp4Ep9Q**; domain=login.abcd.net; path=/; secure,")],
                                                ["Set-Cookie", ("rm_ticket=Uk1Ud.net; expires=Fri, 19-Aug-2016 19:18:48 GMT; path=/; secure,")],
                                                ["Set-Cookie", ("MYPORTAL=bc7; domain=abcd.net; expires=Fri, 19-Aug-2016 19:18:47 GMT; path=/,")],
                                                ["Set-Cookie", ("tls_s_ticket=TFNUAQECAIB_R; domain=abcd.net; path=/; secure,")],
                                                ["Set-Cookie", ("s_ticket=TFNUAQ40-; domain=abcd.net; path=/,"),],
                                                ["Vary", "Accept-Encoding"],
                                                ["Content-Encoding", "gzip"],
                                                ["Keep-Alive", "timeout=5, max=5000"],
                                                ["Connection", "Keep-Alive"],
                                                ["Transfer-Encoding", "chunked"],
                                                ["Content-Type", "text/html"]
                                    ]),
                                    ("<html><head><title>302 Moved Temporarily</title></head>"
"<body bgcolor="#FFFFFF">"
"<p>This document you requested has moved temporarily.</p>"
</p>"
"</body></html>"))
                flow.reply(resp)


        if ("cima.ticket") in flow.request.pretty_url:

            resp = HTTPResponse(
                                 "HTTP/1.1",
                                 302,
                                 "Moved Temporarily",
                                 Headers([
                                            ["Date", "Wed, 21 Jul 2016 23:25:47 GMT"],
                                            ["Server", "Apache"],
                                            ["Cache-Control", "no-cache"],
                                            ["Cache-Control", "no-store"],
                                            ["Pragma", "no-cache"],
                                            ["Expires", "Wed, 31 Dec 1969 23:59:59 GMT"],
                                            ["Location","testme://auth?code=0.ac.V5zGeT&state=xyz"],
                                            ["Set-Cookie","oauth-se!212975827; path=/; HttpOnly"],
                                            ["Set-Cookie",("OAUTHCLIENTS=CgRDU09UEAEYAiKAAYw98*"),],
                                            ["Content-Language", "en-US"],
                                            ["X-Powered-By", "Servlet/2.5 JSP/2.1"],
                                            ["Vary", "Accept-Encoding"],
                                            ["Content-Encoding", "gzip"],
                                            ["Keep-Alive", "timeout=5, max=5000"],
                                            ["Connection", "Keep-Alive"],
                                            ["Transfer-Encoding", "chunked"],
                                            ["Content-Type","text/html"]
                                ]),
                                ("<html><head><title>302 Moved Temporarily</title></head>"
"<body bgcolor="#FFFFFF">"
"<p>This document you requested has moved temporarily.</p>"
"<p>It's now at <a href=\"testme://auth?code=0.ac.V5zGeT&amp;state=xyz\">testme://auth?code=0.ac.V5zGeT&amp;state=xyz</a>.</p>"
"</body></html>"))
            flow.reply(resp)

        if flow.request.pretty_url.endswith("https://login.abcd.net/oauth/token"):

            third_request_file = open('/mock_data_test/access_token.json')
            data = json.load(third_request_file)
            access_token = json.dumps(data)

            resp = HTTPResponse(
                                "HTTP/1.1",
                                200,
                                "OK",
                                Headers([
                                            ["Content-Type","application/x-www-form-urlencoded"],
                                            ["Content-Length","171"]
                                        ]),
                                str(access_token))
            flow.reply(resp)

```

**Comment by joshuabuildsthings** [comment]
@shrikantkaramkar 
Looks like you might have a case that's not triggering any of the "If" statements, so no flow.reply(resp) is getting executed, leading to the mitmproxy Attribute Error you got.

**Comment by shrikantkaramkar** [comment]
I debug the script and found that everything is working fine. 
But still, my application is loading after stopping mitm. It should not require to stop mitm.
Also, on mitm stop I am getting below error.
```
Script error:
__init__() takes at least 6 arguments (5 given)
```

**Comment by joshuabuildsthings** [comment]
@shrikantkaramkar 

For the Script Error we might need more context. As for the mitmproxy stopping; where does it stop? After the first or second redirect? Please try adding the proper Content-Length headers to your 302 redirect responses as well.

**Comment by mhils** [comment]
> AttributeError: 'HTTPResponse' object has no attribute 'response'

This is fixed by #1474, thanks. Regarding the error message, you may just want to update mitmproxy; there's a full traceback in the latest version on master and it seems like you are on master anyways.

**Comment by joshuabuildsthings** [comment]
@mhils - Thanks for clarifying the known bug that was causing at least some of his issue.

I noticed in the current release available at PyPi,  v0.17 that  sometimes mitmproxy hangs after a redirect when the redirect didn't return a Content-Length header. Basically mitmproxy will allow the client to follow the redirect, but doesn't return the final response body until you terminate it. 

I think this was the cause of his initial complaint, but we didn't troubleshoot far enough due to the issue from #1474  popping up.

Any interest in continuing to look into this potential issue?

**Comment by mhils** [comment]
@joshuabuildsthings: Thanks. We're grateful for every bug report that comes with a small, reproducible example on master. If this is the case for the issue you are mentioning, please go ahead and open a new issue! :smiley:
